File: app/src/main/java/org/exhentai/MainActivity.java
package org.exhentai;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;
import android.widget.Toast;

import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import kotlin.text.Regex;


public class MainActivity extends AppCompatActivity {
    private final CookieManager cookieManager = CookieManager.getInstance();
    private ClipboardManager myClipboard;
    private final String loginUrl = "https://forums.e-hentai.org/index.php?act=Login&CODE=00";
    private final String homepageUrl = "https://exhentai.org";
    private final String fullColorUrl = "https://exhentai.org/?f_doujinshi=1&f_manga=1&f_artistcg=1&f_gamecg=1&f_western=0&f_non-h=0&f_imageset=1&f_cosplay=1&f_asianporn=1&f_misc=0&f_search=full+color&f_apply=Apply+Filter";

    private final Pattern galleryRegex = Pattern.compile("https://exhentai\\.org/g/(\\w+)/\\w+");
    private final Pattern galleryPageRegex = Pattern.compile("https://exhentai\\.org/s/\\w+/(\\d+)-(\\d+)");

    private WebView webv;

    public static MainActivity globalActivity;

    public static String exhentaiCookies = "";

    private Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            // Get link-URL.
            String url = (String) msg.getData().get("url");
            addToDownload(url);
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        globalActivity = this;
        setContentView(R.layout.activity_main);
        myClipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        webv = findViewById(R.id.webv);

        String initUrl;
        if (checkCookie()) {
            initUrl = homepageUrl;
            // start download service
            addToDownload(null);
        } else {
            initUrl = loginUrl;
            Toast.makeText(this, R.string.needLogin, Toast.LENGTH_SHORT).show();
        }



        final ProgressBar progressBar = findViewById(R.id.progressBar);
        webv.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                if (newProgress == 100) {
                    progressBar.setVisibility(View.GONE);//加载完网页进度条消失
                } else {
                    progressBar.setVisibility(View.VISIBLE);//开始加载网页时显示进度条
                    progressBar.setProgress(newProgress);//设置进度值
                }
            }
        });
        webv.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                WebView.HitTestResult result = ((WebView) v).getHitTestResult();
                if (result == null) return false;
                int type = result.getType();
                switch (type) {
                    case WebView.HitTestResult.SRC_ANCHOR_TYPE:
                        addToDownload(result.getExtra());
                        break;
                    case WebView.HitTestResult.SRC_IMAGE_ANCHOR_TYPE:
                        Message msg = mHandler.obtainMessage();
                        ((WebView) v).requestFocusNodeHref(msg);
                        break;
                }
                return false;
            }
        });
        webv.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                setTitle(url);
            }
        });
        WebSettings settings = webv.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setSupportZoom(true);
        settings.setBuiltInZoomControls(true);
        settings.setDisplayZoomControls(false);
        settings.setUseWideViewPort(true);
        settings.setLoadWithOverviewMode(true);
        settings.setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);

        initUrl = fullColorUrl;
        webv.loadUrl(initUrl);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addToDownload(webv.getUrl());
            }
        });
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (webv.canGoBack())
                webv.goBack();
            else
                moveTaskToBack(false);
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    Boolean checkCookie() {
        String exhentai = cookieManager.getCookie("exhentai.org");
        if (exhentai != null) {
            String[] cookies = exhentai.split(";");
            if(Pattern.compile("ipb_member_id=\\d{2,}").matcher(exhentai).find()){
                exhentaiCookies = exhentai;
                return true;
            }
            else{
                return false;
            }
        } else {
            return false;
        }
    }

    public void onRefreshClick(MenuItem menuItem) {
        webv.reload();
    }

    public void onFavClick(MenuItem menuItem) {
        webv.loadUrl(fullColorUrl);
    }

    public void onLoadClipboardClick(MenuItem menuItem) {
        ClipData clip = myClipboard.getPrimaryClip();
        if (clip != null && clip.getItemCount() > 0) {
            CharSequence str = clip.getItemAt(0).getText();
            if (str == null) return;
            Log.i("clip", str.toString());
            if (str != null && str.toString().contains("http")) {
                webv.loadUrl(str.toString());
            }
        }
    }

    public void onResetClick(MenuItem menuItem) {
        Intent intent = new Intent(this, HentaiDownloadService.class);
        stopService(intent);
        startService(intent);
    }

    public void onSkipDownloading(MenuItem item) {

    }

    void addToDownload(@Nullable String url) {
        Intent intent = new Intent(this, HentaiDownloadService.class);
        if (url != null) intent.putExtra("url", url);
        startService(intent);
    }

    private Toolbar title;

    void setTitle(String url) {
        if (title == null) title = findViewById(R.id.toolbar);
        Matcher m1 = galleryRegex.matcher(url);
        Matcher m2 = galleryPageRegex.matcher(url);
        if (m1.find()) {
            title.setTitle("ExHentai - " + m1.group(1));
        } else if (m2.find()) {
            title.setTitle("ExHentai - " + m2.group(1) + " - " + m2.group(2));
        } else {
            title.setTitle("ExHentai");
        }
    }
}

File: app/src/main/java/com/yasha/exhentai/MainActivity.kt
package com.yasha.exhentai

import android.app.Activity
import android.content.ClipboardManager
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity;
import android.util.Log
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.webkit.*
import android.widget.Toast

import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import okhttp3.OkHttpClient
import java.io.FileOutputStream
import java.net.URL
import java.util.*
import java.util.concurrent.ArrayBlockingQueue
import kotlin.collections.HashSet

class MainActivity : AppCompatActivity() {
    var cookieManager = CookieManager.getInstance()
    lateinit var myClipboard: ClipboardManager
    val loginUrl = "https://forums.e-hentai.org/index.php?act=Login&CODE=00";
    val homepageUrl = "https://exhentai.org"
    val fullColorUrl = "https://exhentai.org/?f_doujinshi=1&f_manga=1&f_artistcg=1&f_gamecg=1&f_western=0&f_non-h=0&f_imageset=1&f_cosplay=1&f_asianporn=1&f_misc=0&f_search=full+color&f_apply=Apply+Filter"

    val galleryRegex = Regex("https://exhentai\\.org/g/\\w+/\\w+")
    val galleryPageRegex = Regex("https://exhentai\\.org/s/\\w+/\\w+")

    var exhentaiCookie = "ipb_member_id=1601063;ipb_pass_hash=9f4567fb2741f37900a0054d4706a7d2;yay=louder;igneous=ace6704ed;s=7f5a98a89;sk=6a67o8lsurapoheqnzvqwo5g29xu"

    lateinit var downloadManager: HentaiDownloadManager

    companion object {
        var globalActivity: Activity? = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        globalActivity = this
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        myClipboard = this.getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
        var initUrl: String
        if (checkCookie()) {
            initUrl = homepageUrl
        } else {
            initUrl = loginUrl
            Toast.makeText(this, "未设置Cookie，需要登录", Toast.LENGTH_LONG)
        }

        val cookies = HashMap<String, String>()
        for (str in exhentaiCookie.split(";")) {
            val s = str.split("=")
            cookies[s[0]] = s[1]
        }
        downloadManager = HentaiDownloadManager(applicationContext.getExternalFilesDir("").toString(), cookies)

        webv.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, newProgress: Int) {
                if (newProgress == 100) {
                    progressBar.visibility = View.GONE//加载完网页进度条消失
                } else {
                    progressBar.visibility = View.VISIBLE//开始加载网页时显示进度条
                    progressBar.progress = newProgress//设置进度值
                }
            }
        }

        webv.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                Log.i("webv", "finished")
                super.onPageFinished(view, url)
            }
        }
        val settings = webv.settings
        settings.javaScriptEnabled = true
        settings.setSupportZoom(true)
        settings.builtInZoomControls = true
        settings.displayZoomControls = false
        settings.useWideViewPort = true
        settings.loadWithOverviewMode = true
        settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW

        Log.i("UA", settings.userAgentString)

        initUrl = fullColorUrl
        webv.loadUrl(initUrl)

        fab.setOnClickListener { view ->
            //                        Snackbar.make(view, "cookie", Snackbar.LENGTH_LONG)
            //                    .setAction("Action", null).show()
            Log.i("dir", applicationContext.getExternalFilesDir("").toString())
            val url = webv.url
            if (url.contains(galleryRegex)) {
                //取得图集第一页的地址
                webv.evaluateJavascript("(function() { return document.getElementsByClassName('gdtl')[0].firstChild.href; })();") { str ->
                    var page = str?.trim('"')
                    downloadManager.add(page)
                    Log.i("page1", page + "")

                }
            } else if (url.contains(galleryPageRegex)) {
                Log.i("page", url)
                downloadManager.add(url)
            } else {
                Toast.makeText(this, "未能识别地址：地址不是一个图集或图集的一页", Toast.LENGTH_SHORT)
            }
        }
    }

    // 处理返回键
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (webv.canGoBack())
                webv.goBack()
            else
                moveTaskToBack(false)
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    fun checkCookie(): Boolean {
        if (cookieManager.getCookie("e-hentai.org") != null) {
            val hh= cookieManager.getCookie("exhentai.org")
            return true
            val ehentai = "ipb_pass_hash=9f4567fb2741f37900a0054d4706a7d2;ipb_member_id=1601063;ipb_session_id=cdc64a9c2c84b121fe8f203f7fe790e4"
            ehentai.split(';').forEach { str -> cookieManager.setCookie("e-hentai.org", str) }
            if (cookieManager.getCookie("exhentai.org") == null) {
                val time = Date().time / 1000
                val timeStr = time.toString() + "-" + (time + 3 * 60).toString()
                val exhentai = "ipb_member_id=1601063; ipb_pass_hash=9f4567fb2741f37900a0054d4706a7d2; yay=louder; igneous=ace6704ed; s=7f5a98a89; sk=6a67o8lsurapoheqnzvqwo5g29xu; " + timeStr
                exhentai.split(';').forEach { str -> cookieManager.setCookie("exhentai.org", str) }
            }
        }
        return false
    }

    fun onRefreshClick(menuItem: MenuItem) {
        webv.reload()
    }

    fun onFavClick(menuItem: MenuItem) {
        webv.loadUrl(fullColorUrl)
    }

    fun onLoadClipboardClick(menuItem: MenuItem) {
        val clip = myClipboard.primaryClip
        if (clip != null && clip.itemCount > 0) {
            val str = clip.getItemAt(0).text.toString()
            Log.i("clip", str + "")
            if (str != null && str.contains("http")) {
                webv.loadUrl(str)
            }
        }
    }

    fun onResetClick(menuItem: MenuItem) {
        downloadManager.reset()
        val cookies = HashMap<String, String>()
        for (str in exhentaiCookie.split(";")) {
            val s = str.split("=")
            cookies[s[0]] = s[1]
        }
        downloadManager = HentaiDownloadManager(applicationContext.getExternalFilesDir("").toString(), cookies)    }
}
